sentrework.tex

```diff
@@ -0,0 +1,230 @@
+their common type.
+
+\pnum
+Returning to the example above, the consequences of the
+\tcode{EqualityComparable} requirement are clear: since \tcode{i+2 == S\{2\}}
+and \tcode{i+2 == S\{3\}}, \tcode{S\{2\}} and \tcode{S\{3\}} must be equal. If
+they are equal, then the ranges \range{i+0}{S\{2\}} and \range{i+0}{S\{3\}} must
+also denote the same range due to the equality preservation of the expressions
+involved. This is, however, not the case since \tcode{*(i+0) == 2} is greater
+than or equal to \tcode{2} but not greater than or equal to \tcode{3}.
+\tcode{EqualityComparable<S,int*>()} and therefore \tcode{Sentinel<S,int*>()}
+are not satisfied; \tcode{S} is not a valid sentinel for \tcode{int*}.
+
+\pnum
+Clearly there is a large class of useful sentinels that the N4560 model does
+not admit: stateful sentinels which might compare equal to the same iterator
+when they have differing state. N4560 constrains iterators and sentinels to
+both represent the abstract notion of ``position,'' which is not a good fit
+for stateful sentinels.
+
+\rSec1[sentrework.approach]{Fixing the problem}
+
+\pnum
+It seems odd to claim that the sentinel semantics are broken when there are two
+working implementations of the TS~(\cite{range-v3}, \cite{cmcstl2}) that haven't
+had any issues due to this semantic inconsistency. The clear implication is that
+the semantics of the N4560 model are not an exact match for the requirements of
+the algorithms. The semantics are likely overconstrained and could be relaxed
+without losing the properties the algorithms need to support equational
+reasoning.
+
+\rSec2[sentrework.approach.alg]{Look at the Algorithms}
+
+\pnum
+As is often the case, a problem in the semantics is best fixed by examining
+the algorithms and determining what behaviors they require to ensure their
+postconditions. The properties the algorithms require for sentinels are (where
+\tcode{s} denotes a sentinel value and \tcode{i} an iterator value):
+\begin{itemize}
+\item \tcode{i == s}, \tcode{s == i}, \tcode{i != s}, and \tcode{s != i} are
+      equality-preserving expressions with the same definition space.
+\item Symmetry: \tcode{bool(i == s) == bool(s == i)} and
+      \tcode{bool(i != s) == bool(s != i)}.
+\item Complement: \tcode{bool(i != s) == !bool(i == s)}.
+\item \tcode{i == s} is well-defined when \range{i}{s} denotes a range.
+\end{itemize}
+
+\pnum
+The first three properties aren't specific to iterators and sentinels; they are
+general characteristics of the operators \tcode{==} and \tcode{!=} that are
+intuitively expected to hold whenever they are defined. As such, the
+\tcode{EqualityComparable} concepts should require them to hold and this
+particular set of properties can then be presented as a relaxation of
+\tcode{EqualityComparable<T,U>()}. The current design does exactly that,
+defining the concept
+\tcode{Bike\-shed\-Equal\-ity\-Comp\-arable}~(\ref{concepts.lib.compare.equalitycomparable})
```

Comment:
We'll go with "Weak" until something better comes along. Once I started laughing at the double meaning of `UncommonlyEqualityComparable` I was incapable of thinking of a reasonable name.